runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/FanOutStreamingEngineWorkerHarness.java

```diff
@@ -256,58 +242,132 @@ private GetDataStream getGlobalDataStream(String globalDataKey) {
     Preconditions.checkState(started, "StreamingEngineClient never started.");
     getWorkerMetadataStream.get().halfClose();
-    getWorkBudgetRefresher.stop();
-    newWorkerMetadataPublisher.shutdownNow();
-    newWorkerMetadataConsumer.shutdownNow();
+    workerMetadataConsumer.shutdownNow();
     channelCachingStubFactory.shutdown();
   }
 
-  /**
-   * {@link java.util.function.Consumer<WindmillEndpoints>} used to update {@link #connections} on
-   * new backend worker metadata.
-   */
+  @SuppressWarnings("methodref.receiver.bound")
+  private Supplier<GetWorkerMetadataStream> createGetWorkerMetadataStream(
+      @UnderInitialization FanOutStreamingEngineWorkerHarness this) {
+    // Checker Framework complains about reference to "this" in the constructor since the instance
+    // is "UnderInitialization" here, which we pass as a lambda to GetWorkerMetadataStream for
+    // processing new worker metadata. Supplier.get() is only called in start(), after we have
+    // constructed the FanOutStreamingEngineWorkerHarness.
+    return () ->
+        checkNotNull(streamFactory)
+            .createGetWorkerMetadataStream(
+                checkNotNull(dispatcherClient).getWindmillMetadataServiceStubBlocking(),
+                checkNotNull(getWorkerMetadataThrottleTimer),
+                this::consumeWorkerMetadata);
+  }
+
+  private void consumeWorkerMetadata(WindmillEndpoints windmillEndpoints) {
+    synchronized (metadataLock) {
+      // Only process versions greater than what we currently have to prevent double processing of
+      // metadata. workerMetadataConsumer is single-threaded so we maintain ordering.
+      if (windmillEndpoints.version() > pendingMetadataVersion) {
+        pendingMetadataVersion = windmillEndpoints.version();
+        workerMetadataConsumer.execute(() -> consumeWindmillWorkerEndpoints(windmillEndpoints));
+      }
+    }
+  }
+
   private synchronized void consumeWindmillWorkerEndpoints(WindmillEndpoints newWindmillEndpoints) {
-    isBudgetRefreshPaused.set(true);
+    // Since this is run on a single threaded executor, multiple versions of the metadata maybe
+    // queued up while a previous version of the windmillEndpoints were being consumed. Only consume
+    // the endpoints if they are the most current version.
+    synchronized (metadataLock) {
+      if (newWindmillEndpoints.version() < pendingMetadataVersion) {
+        return;
+      }
+    }
+
+    long previousMetadataVersion = activeMetadataVersion;
     LOG.info("Consuming new windmill endpoints: {}", newWindmillEndpoints);
     ImmutableMap<Endpoint, WindmillConnection> newWindmillConnections =
         createNewWindmillConnections(newWindmillEndpoints.windmillEndpoints());
-
+    closeStaleStreams(newWindmillConnections.values(), connections.get().windmillStreams());
+    ImmutableMap<WindmillConnection, WindmillStreamSender> newStreams =
+        createAndStartNewStreams(newWindmillConnections.values()).join();
```

Comment:
we only want 1 set of streams to be active at a time. since when we chose what streams to start/close we snapshot the current connections state.  I think we want the start operation to be atomic?

we don't block on closing the streams just starting them